### PR TITLE
[o1vm] Dynamic Selectors for MIPS

### PR DESCRIPTION
**Pull request by Fizzixnerd**

### CODE REVIEW

**Comment by Fizzixnerd** [comment]
Should be ready for review;

I will add my own review covering stuff I'm not sure of.

File: o1vm/src/interpreters/mips/witness.rs
Comment (dannywillems) [inline o1vm/src/interpreters/mips/witness.rs:86]: I suppose you can use a boolean instead of a field element.

Comment (dannywillems) [reply o1vm/src/interpreters/mips/witness.rs:86]: You can also keep simply the integer value of the current selector.

File: o1vm/src/interpreters/mips/interpreter.rs
Comment (dannywillems) [inline o1vm/src/interpreters/mips/interpreter.rs:995]: I would not add this in the interpreter, but instead outside, when getting the constraints.

File: o1vm/src/interpreters/mips/interpreter.rs
Comment (dannywillems) [inline o1vm/src/interpreters/mips/interpreter.rs:1001]: I would not add this in the interpreter, but instead outside, when getting the constraints.

File: o1vm/src/interpreters/mips/tests_helpers.rs
Comment (dannywillems) [inline o1vm/src/interpreters/mips/tests_helpers.rs:94]: It might be ok for a first version, but I would simply keep track of the actual integer representing the instruction, and in the `main.rs` file, I would build the actual vectors of evaluation points.

**Review by dannywillems — COMMENTED** [review]
Look at how it was done initially: https://github.com/o1-labs/proof-systems/blob/5b687813895fef22a5fc199103b70c3fb0d2360e/optimism/src/main.rs#L84.

I would go with this kind of version.

File: o1vm/src/interpreters/mips/constraints.rs
Comment (marcbeunardeau88) [reply o1vm/src/interpreters/mips/constraints.rs:73]: It's not correct since it's empty
Or you mean you want to do this in a follow up ?

**Comment by Fizzixnerd** [comment]
@dannywillems I deleted the code I added to the witness, as it's not clear it's needed now.  Will add back later if it is, but for now this just adds the selectors to the proof inputs.

File: o1vm/src/interpreters/mips/constraints.rs
Comment (Fizzixnerd) [inline o1vm/src/interpreters/mips/constraints.rs:77]: Is there a less verbose way of writing this?  it's just `-1`.

File: o1vm/src/interpreters/mips/constraints.rs
Comment (Fizzixnerd) [inline o1vm/src/interpreters/mips/constraints.rs:81]: Add that they are all either zero or one

File: o1vm/src/interpreters/mips/constraints.rs
Comment (Fizzixnerd) [inline o1vm/src/interpreters/mips/constraints.rs:83]: Because `acc` starts at -1, this is zero if only one is one.

File: o1vm/src/interpreters/mips/witness.rs
Comment (dannywillems) [inline o1vm/src/interpreters/mips/witness.rs:147]: This is not correct:
`variable` must return the actual value saved at the position `column`.
If the column is a `SCRATCH_STATE`, we get the value from `self.scratch_state`. If the value is a `Selector(idx)`, you should return 1 if it is the index of the selector for the current instruction, otherwise 0.

File: o1vm/src/pickles/main.rs
Comment (dannywillems) [inline o1vm/src/pickles/main.rs:92]: Note: you can simply use `Fp::from(instr as u64)`

File: o1vm/src/interpreters/mips/column.rs
Comment (dannywillems) [inline o1vm/src/interpreters/mips/column.rs:177]: This is not correct. You get an underflow.

File: o1vm/src/interpreters/mips/constraints.rs
Comment (dannywillems) [inline o1vm/src/interpreters/mips/constraints.rs:78]: You can use `Self::Variable::one()` and import `ark_ff::One`.

**Review by dannywillems — CHANGES_REQUESTED** [review]
Running with the nodes give errors:
```
[2024-10-02T17:29:43Z INFO  o1vm::interpreters::mips::witness] processing step=0 pc=00000ae88c insn=000802ba1f ips=0.00 pages=8598 mem=33.6 MiB name=_rt0_mips_linux
[2024-10-02T17:29:46Z INFO  o1vm::interpreters::mips::witness] Snapshot state in snapshot-state-0.json, step 0
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: TryFromIntError(())', o1vm/src/pickles/main.rs:93:81
```

File: o1vm/src/interpreters/mips/column.rs
Comment (dannywillems) [reply o1vm/src/interpreters/mips/column.rs:177]: This is correct. `usize::from(self)` is not the conversion from `EnumIter` into a `usize`, it is actually an implementation from Instruction to `usize` defined above. Sorry.

**Comment by dannywillems** [comment]
See https://github.com/o1-labs/proof-systems/pull/2649

**Review by dannywillems — APPROVED** [review]
As https://github.com/o1-labs/proof-systems/pull/2649/ has been approved, and has been built on top of this PR, I would approve this one, and merge right after 2649.

**Comment by dannywillems** [comment]
Taking care of fixing conflicts

File: o1vm/src/interpreters/mips/constraints.rs
Comment (marcbeunardeau88) [reply o1vm/src/interpreters/mips/constraints.rs]: There is a From<u64> for expr

**Pull request merged**
